### PR TITLE
User specs

### LINKED ISSUE #12: OBJECT MODEL: Add user model specs
**ACCEPTANCE**
- [x] Basic user object behavior is tested

Possible inspiration (the user model may be slightly different due to differences between Tufts and OHSU requirements)
https://github.com/curationexperts/epigaea/blob/master/spec/models/user_spec.rb

### PR DESCRIPTION
**Pull request by bess**
Fixes #12

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  it '#to_s yeilds the #display_name' do
+    expect(user.to_s).to eq user.display_name
+  end
+
+  describe 'roles' do
+    it 'is emplty for a new user' do
+      new_user = create(:user)
+      expect(new_user.roles).to be_empty
```
Comment (no-reply) [inline spec/models/user_spec.rb]: Could we do `expect(plain_user.roles).to be_empty` instead? This has the advantage of ensuring that `plain_user` doesn't accidentally become non-plain in the future, invalidating other tests.

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    it "#add_role uses an existing role if it exists" do
+      role = Role.create(name: 'existing_role')
+      plain_user.add_role('existing_role')
+      expect(plain_user.roles.to_a).to include role
+    end
+
+    it "#remove_role removes a named role" do
+      new_user = create(:user)
+      new_user.add_role('new_role')
```
Comment (no-reply) [inline spec/models/user_spec.rb]: `plain_user` again?

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      expect { new_user.remove_role('new_role') }.to change { new_user.roles.count }.from(1).to(0)
+    end
+
+    it "#remove_role does nothing on non-existant names" do
+      existing_roles = plain_user.roles
+      plain_user.remove_role('nonexistent_role_name')
```
Comment (no-reply) [inline spec/models/user_spec.rb]: Do you have an opinion about using `.not_to change` here?

```ruby
expect { plain_user.remove_role('nonexistent_role_name') }
  .not_to change { plain_user.roles.to_a }
```

This feels a little bit more declarative/explicit to me.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    end
+  end
+
+  describe '#admin?' do
+    it 'is false for regular user' do
+      expect(user).not_to be_admin
+    end
+
+    it 'is true when a user has the "admin" role' do
+      admin_user = create(:user)
```
Comment (no-reply) [inline spec/models/user_spec.rb]: Should we use the admin factory? Maybe with `let(:admin_user) { create(:admin) }`?

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      expect(plain_user.roles.inspect).to match(/some_role/)
+    end
+
+    it "#add_role creates a new role if needed" do
+      expect { plain_user.add_role('named_role') }.to change { Role.count }.by(1)
+    end
+
+    it "#add_role uses an existing role if it exists" do
+      role = Role.create(name: 'existing_role')
+      plain_user.add_role('existing_role')
```
Comment (no-reply) [inline spec/models/user_spec.rb]: `.to change` here? This could pass if `'existing_role'` was somehow added in setup.

**Review by no-reply — COMMENTED** [review]
This looks good and I think it could be merged as is.

I have a few minor spec refactor comments that I'll defer to you about whether to address or not.

Can we squash the rubocop fixup commit back into the main work? If you push up the squash, I'll approve and merge. :rocket:

File: spec/models/user_spec.rb (hunk 1 above)
Comment (bess) [reply spec/models/user_spec.rb]: 👍

File: spec/models/user_spec.rb (hunk 5 above)
Comment (bess) [reply spec/models/user_spec.rb]: 👍

File: spec/models/user_spec.rb (hunk 2 above)
Comment (bess) [reply spec/models/user_spec.rb]: 👍

File: spec/models/user_spec.rb (hunk 3 above)
Comment (bess) [reply spec/models/user_spec.rb]: 👍

**Comment by bess** [comment]
@no-reply Refactored with your improvements. Take a look, please?

File: spec/features/create_etd_spec.rb [hunk 6]
```diff
@@ -49,4 +60,30 @@
...
     end
   end
+
+  # Non admin users should not be able to create a new work or to edit any works
+
+  context "a non-admin user" do
+    before do
+      login_as user
+    end
+    scenario "can see a work" do
```
Comment (no-reply) [inline spec/features/create_etd_spec.rb]: Is this already tested in https://github.com/curationexperts/mahonia/blob/496863d28d768115d7db004919bd034cb922b06d/spec/features/access_etd_spec.rb ?

**Comment by mark-dce** [comment]
connected to #12

File: spec/features/create_etd_spec.rb (hunk 6 above)
Comment (bess) [reply spec/features/create_etd_spec.rb]: Okay, redundant test removed

**Review by no-reply — APPROVED** [review]
Thanks for your patience on cycling this, @bess!

**Pull request merged**